tests/test_mpls.py

```diff
@@ -0,0 +1,1049 @@
+        self.create_mpls_intf("Ethernet4")
+        self.create_mpls_intf("Ethernet8")
+
+        # set ip address
+        self.add_ip_address("Ethernet0", "10.0.0.0/31")
+        self.add_ip_address("Ethernet4", "10.0.0.2/31")
+        self.add_ip_address("Ethernet4", "10.0.0.4/31")
+
+        # bring up interface
+        self.set_admin_status("Ethernet0", "up")
+        self.set_admin_status("Ethernet4", "up")
+        self.set_admin_status("Ethernet8", "up")
+
+        # set ip address and default route
+        dvs.servers[0].runcmd("ip address add 10.0.0.1/31 dev eth0")
+        dvs.servers[0].runcmd("ip route add default via 10.0.0.0")
+
+        dvs.servers[1].runcmd("ip address add 10.0.0.3/31 dev eth0")
+        dvs.servers[1].runcmd("ip route add default via 10.0.0.2")
+
+        dvs.servers[2].runcmd("ip address add 10.0.0.5/31 dev eth0")
+        dvs.servers[2].runcmd("ip route add default via 10.0.0.4")
+
+        # get neighbor and arp entry
+        dvs.servers[0].runcmd("ping -c 1 10.0.0.3")
+        dvs.servers[2].runcmd("ping -c 1 10.0.0.3")
+
+        # add route entry
+        dvs.runcmd("ip -f mpls route add 200 nexthop via inet 10.0.0.1 dev Ethernet0 nexthop via inet 10.0.0.5 dev Ethernet8")
+
+        # check application database
+        self.pdb.wait_for_entry("LABEL_ROUTE_TABLE", "200")
+
+        # check ASIC inseg database
+        self.check_inseg_entries(["200"])
+
+        # remove route entry
+        dvs.runcmd("ip -f mpls route del 200 nexthop via inet 10.0.0.1 dev Ethernet0 nexthop via inet 10.0.0.5 dev Ethernet8")
+
+        # check application database
+        self.pdb.wait_for_deleted_entry("LABEL_ROUTE_TABLE", "200")
+
+        # check ASIC inseg database
+        self.check_deleted_inseg_entries(["200"])
+
+        # remove ip address
+        self.remove_ip_address("Ethernet0", "10.0.0.0/31")
+        self.remove_ip_address("Ethernet4", "10.0.0.2/31")
+        self.remove_ip_address("Ethernet8", "10.0.0.4/31")
+
+        # remove mpls interface
+        self.remove_mpls_intf("Ethernet0")
+        self.remove_mpls_intf("Ethernet4")
+        self.remove_mpls_intf("Ethernet8")
+
+        self.set_admin_status("Ethernet0", "down")
+        self.set_admin_status("Ethernet4", "down")
+        self.set_admin_status("Ethernet8", "down")
+
+        # remove ip address and default route
+        dvs.servers[0].runcmd("ip route del default dev eth0")
+        dvs.servers[0].runcmd("ip address del 10.0.0.1/31 dev eth0")
+
+        dvs.servers[1].runcmd("ip route del default dev eth0")
+        dvs.servers[1].runcmd("ip address del 10.0.0.3/31 dev eth0")
+
+        dvs.servers[2].runcmd("ip route del default dev eth0")
+        dvs.servers[2].runcmd("ip address del 10.0.0.5/31 dev eth0")
+
+        self.clear_mpls(dvs)
+
+class TestMplsRouteAppDb(TestMplsBase):
+    """ Functionality tests for mpls triggered from APP_DB """
+    def test_RouteAddRemoveMplsPushRoute(self, dvs, testlog):
```

Comment:
suggest separating out the tests. It's easier to debug failed tests when testing one specific part. You can have 'setup' and 'teardown' APIs to do common stuff.

Reply:
@smahesh Not sure what you mean by separating out the tests.  I based the current test_mpls.py hierarchy on the existing testcases in test_route.py, so I would assume that the structure of test_mpls.py is also fine.  Please be more specific about what you are requesting here.

Reply:
I meant "add" can be one test, "delete" can be another test. It's alright if the test case is too small add/del can be in the same test case. Once the common code is consolidated the test case might end up being few lines.

Reply:
@smaheshm please re-review code with common code consolidation.